### PR TITLE
docs: update Consul and Vault integration

### LINKED ISSUE #19456: Documentation: dead links/incorrect links
In the documentation for vault integration (https://developer.hashicorp.com/nomad/docs/integrations/vault-integration#migrating-to-using-workload-identity-with-vault) there are 2 links referenced that go to the wrong section and basically do not work: 

> Enable [vault.default_identity](https://developer.hashicorp.com/nomad/docs/configuration/consul#default_identity) blocks in your Nomad server agent configurations.
>
> (Optionally) Add [vault.role](https://developer.hashicorp.com/nomad/docs/configuration/consul#role) fields to any Nomad jobs that will not use the default role.

The "consul" section does not have said anchors, and since we're talking about vault integration I'm pretty sure they need to be sent to the vault block(s) docs.

### PR DESCRIPTION
**Pull request by lgfa29**
Expand the Consul and Vault integration docs to explain how the new workload identity flow works.

Preview of pages with the biggest changes:
* [Consul Integration - ACL](https://nomad-git-docs-consul-vault-integrations-hashicorp.vercel.app/nomad/docs/integrations/consul/acl)
* [Vault Integration - ACL](https://nomad-git-docs-consul-vault-integrations-hashicorp.vercel.app/nomad/docs/integrations/vault/acl)
* [Concepts - Workload Identity](https://nomad-git-docs-consul-vault-integrations-hashicorp.vercel.app/nomad/docs/concepts/workload-identity)
* [Jobspec - `identity`](https://nomad-git-docs-consul-vault-integrations-hashicorp.vercel.app/nomad/docs/job-specification/identity)

Closes #19456

### CODE REVIEW

**Review by tgross — COMMENTED** [review]
Left a few comments but I'm not done yet :grinning:

File: website/content/docs/integrations/consul-integration.mdx [hunk 1]
```diff
@@ -93,16 +262,45 @@ remaining fields are required to match those shown here.
...
 
-</Tab>
-<Tab heading="Consul Enterprise">
+</CodeBlockConfig>
+
+The `JWKSURL` address must be reachable by all Consul servers and should
+resolve to multiple Nomad agents to avoid a single point of failure. Both Nomad
+servers and clients are able to handle this request.
+
+[![Consul Auth Method][img_consul_auth_method]][img_consul_auth_method]
```
Comment (tgross) [inline website/content/docs/integrations/consul-integration.mdx:271]: This doc has a "Server" block that includes both Nomad and Consul. Maybe we should clarify this to be "Nomad Server"?

File: website/content/docs/integrations/consul-integration.mdx [hunk 2]
```diff
@@ -93,16 +262,45 @@ remaining fields are required to match those shown here.
...
+
+When Nomad is configured to use mTLS, [`tls.verify_https_client`][] must be set
+to `false` since it is not possible to provide client certificates to the
+Consul auth method. Alternatively, you may expose Nomad's JWKS URL from a proxy
+or load balancer that terminates the TLS connection.
+
+If the Consul servers are not able to reach Nomad's JWKS URL, the public keys
+may be read from Nomad's `/.well-known/jwks.json` and provided to the auth
+method directly using the [`JWTValidationPubKeys`][] parameter. The keys must
```
Comment (tgross) [inline website/content/docs/integrations/consul-integration.mdx:280]: The passive voice makes this ambiguous. I would word this as "you may read the public keys..." so that it's clear this is something that you, the cluster admin, needs to do manually.

(Same for this paragraph's twin in the Vault doc)

File: website/content/docs/integrations/consul-integration.mdx [hunk 3]
```diff
@@ -93,16 +262,45 @@ remaining fields are required to match those shown here.
...
+be converted from JWKS to PEM format and **are rotated periodically**, so this
+process must be done continually. The rotation frequency is controlled by the
+[`server.root_key_rotation_threshold`][] configuration of the Nomad servers.
+
+When an allocation that needs access access to Consul starts, the Nomad client
+running it exchanges the Nomad workload identities for Consul from tasks and
+services for Consul ACL tokens.
```
Comment (tgross) [inline website/content/docs/integrations/consul-integration.mdx:287]: It's clear from context that these are the WIs for Consul services and leaving it makes this sentence a little hard to parse (with "for Consul" being in there twice).

```suggestion
When an allocation that needs access access to Consul starts, the Nomad client
running it exchanges the Nomad workload identities for tasks and services for 
Consul ACL tokens.
```

File: website/content/docs/integrations/consul-integration.mdx [hunk 4]
```diff
@@ -111,134 +309,174 @@ remaining fields are required to match those shown here.
...
+claim. The [`-selector`][] flag ensures this binding rule only applies to
+workload identities for services.
 
-</Note>
+The second binding rule associates the Consul ACL token with an [ACL
+role][consul_acl_role], which is a collection of [ACL
+policies][consul_acl_policy] that define what the token is authorized to do.
+This binding rule is applied to Nomad workload identities for tasks to access
+information from Consul. The exact ACL policy rules will depend on the level of
+access required by the tasks.
```
Comment (tgross) [inline website/content/docs/integrations/consul-integration.mdx:369]: Maybe add a parenthetical here like "(typically, accessing service addresses and KV with `template` blocks)"?

File: website/content/docs/integrations/vault-integration.mdx [hunk 5]
```diff
@@ -22,134 +22,356 @@ install Vault separately from Nomad. Nomad does not run Vault for you.
...
+the level of access required by the tasks.
 
-Next, write a Vault policy file using that accessor. Note that `auth_jwt_3a9350fe` here
- is the accessor shown above and you should replace it with
-the results from your own Vault cluster.
+<CodeBlockConfig highlight="1,5,9" filename="acl-policy.hcl">
 
 ```hcl
-path "secret/data/{{identity.entity.aliases.auth_jwt_3a9350fe.metadata.nomad_namespace}}/{{identity.entity.aliases.auth_jwt_3a9350fe.metadata.nomad_job_id}}/*" {
+path "kv/data/{{identity.entity.aliases.auth_jwt_d34481ad.metadata.nomad_namespace}}/{{identity.entity.aliases.auth_jwt_d34481ad.metadata.nomad_job_id}}/*" {
```
Comment (tgross) [inline website/content/docs/integrations/vault-integration.mdx:333]: The default path for KV is `secret/`

File: website/content/docs/concepts/workload-identity.mdx [hunk 6]
```diff
@@ -6,11 +6,16 @@ description: Learn about Nomad's workload identity feature
...
 # Workload Identity
 
-Every workload running in Nomad is given an identity. When an [allocation][] is
-accepted by the [plan applier][], the leader generates a Workload Identity for
-each task in the allocation. This workload identity is a [JSON Web Token
-(JWT)][] that has been signed by the leader's keyring. The workload identity
-includes the following identity claims:
+Every workload running in Nomad is given a default identity. When an
+[allocation][] is accepted by the [plan applier][], the leader generates a
+Workload Identity for each task in the allocation. This workload identity is a
```
Comment (Juanadelacuesta) [inline website/content/docs/concepts/workload-identity.mdx:11]: Can one allocation have multiple tasks?

File: website/content/docs/integrations/vault-integration.mdx [hunk 7]
```diff
@@ -22,134 +22,356 @@ install Vault separately from Nomad. Nomad does not run Vault for you.
...
 
-## Using Nomad Workload Identity with Vault
+### Nomad Workload Identities
 
 Starting in Nomad 1.7, Nomad clients will use a task's [Workload Identity][] to
 authenticate to Vault and obtain a Vault token specific to the task.
 
-### Configuring Vault Authentication
+By default, Nomad only generates a workload identity for tasks that is used to
+access Nomad itself. To access Vault, jobs must define additional workload
```
Comment (Juanadelacuesta) [inline website/content/docs/integrations/vault-integration.mdx:41]: This line is a bit hard to follow, Nomad only generates identities for workloads that need to access some other Nomad resources?

File: website/content/docs/integrations/vault-integration.mdx [hunk 8]
```diff
@@ -22,134 +22,356 @@ install Vault separately from Nomad. Nomad does not run Vault for you.
...
-$ vault auth enable jwt
+    task "mongo" {
+      driver = "docker"
+
+      config {
+        image = "mongo:7"
+        ports = ["db"]
+      }
+
+      vault {}
```
Comment (Juanadelacuesta) [inline website/content/docs/integrations/vault-integration.mdx:78]: The text in line 47 mentions a `vault.cluster`, but the `vault` block here is empty.

File: website/content/docs/integrations/consul-integration.mdx [hunk 9]
```diff
@@ -50,38 +50,207 @@ Consul data, Vault secrets, or just general configurations within a Nomad task.
...
+
+By default, Nomad does not generate workload identities for services, and tasks
+only have a default identity that is used to access Nomad itself. To access
+Consul, jobs must define additional workload identities suitable for Consul
+access.
+
+These additional identities can be defined as an [`identity`][] block inside
+`task` or `service` blocks. When specified inside a `task`, the
+[`identity.name`][] value must follow the pattern `consul_<cluster name>`,
+where `<cluster name>` is the value for [`consul.cluster`][] of the `consul`
```
Comment (Juanadelacuesta) [inline website/content/docs/integrations/consul-integration.mdx:123]: Similar to the vault example, it is not very clear to mention the `consul` block requiring a name and then it not being necessary and not shown on the code

**Review by Juanadelacuesta — APPROVED** [review]
Good job!!! it is easy to follow and understand, despite being a hairy subject. 
Just a couple of minor comments.

File: website/content/docs/integrations/consul-integration.mdx [hunk 10]
```diff
@@ -50,38 +50,207 @@ Consul data, Vault secrets, or just general configurations within a Nomad task.
...
-remaining fields are required to match those shown here.
+Starting in Nomad 1.7, Nomad clients can use a task or service's [Workload
+Identity][nomad_wid] to authenticate to Consul and obtain an ACL token specific
+to the service or task. When using Nomad workload identities, you no longer
+need to pass in a Consul ACL token to submit a job.
+
+By default, Nomad does not generate workload identities for services, and tasks
+only have a default identity that is used to access Nomad itself. To access
+Consul, jobs must define additional workload identities suitable for Consul
+access.
```
[the remark this answers is not in the document]
Comment (lgfa29) [reply website/content/docs/integrations/consul-integration.mdx:118]: I wanted to document how manual identities work, but since we probably want to nudge people towards automatic identity injection, I should probably keep the focus on this page about that?

Then I can move this part somewhere else (maybe new sections in `Concepts/Workload Identity`) and reference there for more details on custom identities.

File: website/content/docs/integrations/consul-integration.mdx [hunk 11]
```diff
@@ -50,38 +50,207 @@ Consul data, Vault secrets, or just general configurations within a Nomad task.
...
+URL][nomad_jwks_url]. Consul servers call this URL to retrieve the public keys
+Nomad used to sign the workload identities so they can validate their origin
+and confirm that they were created by Nomad.
+
+<CodeBlockConfig highlight="2" filename="auth-method.json">
 
 ```json
 {
-  "JWKSURL": "https://nomad.example.com:4646/.well-known/jwks.json",
+  "JWKSURL": "http://nomad.example.com:4646/.well-known/jwks.json",
```
[the remark this answers is not in the document]
Comment (lgfa29) [reply website/content/docs/integrations/consul-integration.mdx:253]: Ah yeah. I just copied the examples from the tutorial 😅

File: website/content/docs/concepts/workload-identity.mdx (hunk 6 above)
Comment (lgfa29) [reply website/content/docs/concepts/workload-identity.mdx:11]: Yup! A group generates one or more allocations (depending on its `count` value), and since each `group` can have one or more tasks, the allocation can also have multiple tasks.

File: website/content/docs/integrations/vault-integration.mdx (hunk 7 above)
Comment (lgfa29) [reply website/content/docs/integrations/vault-integration.mdx:41]: Yup! For example, when reading [Nomad variables from a `template`](https://developer.hashicorp.com/nomad/docs/job-specification/template#nomad-integration) the task needs to access the Nomad API, so it uses this identity for authentication.

File: website/content/docs/integrations/vault-integration.mdx (hunk 8 above)
Comment (lgfa29) [reply website/content/docs/integrations/vault-integration.mdx:78]: Yeah, `cluster` is optional and only available in Nomad Enterprise.

File: website/content/docs/integrations/consul-integration.mdx (hunk 9 above)
Comment (lgfa29) [reply website/content/docs/integrations/consul-integration.mdx:123]: Ah good point, we actually _should_ add it to the example, even if not required, to nudge folks to start using (ref. https://github.com/hashicorp/nomad/issues/19390)

File: website/content/docs/integrations/consul-integration.mdx (hunk 1 above)
Comment (lgfa29) [reply website/content/docs/integrations/consul-integration.mdx:271]: Good point. Added some extra label to each box to indicate what they mean. Not sure if this is the best approach so I will gather some feedback 😅 
image

File: website/content/docs/integrations/consul-integration.mdx (hunk 3 above)
Comment (lgfa29) [reply website/content/docs/integrations/consul-integration.mdx:287]: I also had `access` twice in there 😅

File: website/content/docs/integrations/consul-integration.mdx (hunk 10 above)
[the remark this answers is not in the document]
Comment (lgfa29) [reply website/content/docs/integrations/consul-integration.mdx]: After looking at it again, I thought it would be better to description of custom identity name in the `identity` block page because that's what being modified:
https://nomad-git-docs-consul-vault-integrations-hashicorp.vercel.app/nomad/docs/job-specification/identity#workload-identities-for-consul

File: website/content/docs/integrations/consul-integration.mdx (hunk 1 above)
Comment (lgfa29) [reply website/content/docs/integrations/consul-integration.mdx]: I got professional help and it works much better now 😄 
image

File: website/content/docs/integrations/consul-integration.mdx (hunk 2 above)
Comment (lgfa29) [reply website/content/docs/integrations/consul-integration.mdx]: I tried to keep this short since it's in the middle of a long text, but it's important information to have, so I moved this and the mTLS discussion to their own section:
https://nomad-git-docs-consul-vault-integrations-hashicorp.vercel.app/nomad/docs/integrations/consul/acl#important-considerations-about-the-jwks-url

This allows us to expand them as necessary.

File: website/content/docs/integrations/vault-integration.mdx (hunk 7 above)
Comment (lgfa29) [reply website/content/docs/integrations/vault-integration.mdx]: Expanded to this:

> By default, Nomad does not generate workload identities for services, and tasks only receive an identity that is used to access data from Nomad itself, such as for reading [Variables](https://nomad-git-docs-consul-vault-integrations-hashicorp.vercel.app/nomad/docs/concepts/variables) from a [template](https://nomad-git-docs-consul-vault-integrations-hashicorp.vercel.app/nomad/docs/job-specification/template) block.

File: website/content/docs/integrations/consul-integration.mdx (hunk 9 above)
Comment (lgfa29) [reply website/content/docs/integrations/consul-integration.mdx]: Added `consul` block on CE example as well:
https://nomad-git-docs-consul-vault-integrations-hashicorp.vercel.app/nomad/docs/job-specification/identity#workload-identities-for-consul

**Comment by lgfa29** [comment]
I ended up restructuring the whole `Integrations` section to pull the ACL part onto their pages since this is really important and extensive information. Cramming everything into a single page was making it hard to read and creating really deeply nested sections.

It also gives us room to expand on more detailed integration docs, with specific feature descriptions and how-tos (like for Consul Service Mesh now).

File: website/content/docs/job-specification/identity.mdx [hunk 12]
```diff
@@ -10,7 +10,9 @@ description: |-
 
 <Placement
   groups={[
+    ['job', 'group', 'service', 'identity'],
     ['job', 'group', 'task', 'identity'],
+    ['job', 'group', 'task', 'service', 'identity'],
   ]}
```
Comment (lgfa29) [inline website/content/docs/job-specification/identity.mdx:16]: I'm not sure about the ordering here. `job -> group -> task -> identity` is arguably  the most important placement, but moving it first make things look out of order.

File: website/content/docs/integrations/vault/acl.mdx [hunk 13]
```diff
@@ -140,14 +260,66 @@ path "secret/metadata/*" {
...
+#### Mutual TLS in Nomad
+
+It is highly recommended to use [mutual TLS][tutorial_mtls] in production
+deployments of Nomad. With mTLS enabled, the [`tls.verify_https_client`][]
+configuration must be set to `false` since it is not possible to provide client
+certificates to the Vault auth method.
+
+Alternatively, you may expose Nomad's JWKS URL from a proxy or a load balancer
+that handles the mutual TLS connection to Nomad and exposes the JWKS URL
+endpoint over standard TLS.
```
Comment (lgfa29) [inline website/content/docs/integrations/vault/acl.mdx:294]: Is there a name for non-mutual TLS? I used "standard TLS" but I'm not sure if that's right 😅

**Review by tgross — APPROVED** [review]
LGTM! I've left some comments but nothing super critical. Nice work on this, @lgfa29!

**Comment by lgfa29** [comment]
We should be adding redirects to our repo again, so I pushed a commit with them.

https://developer.hashicorp.com/ is also not able to handle the static file URLs we used to download Vault config files. It appends a random string to all static assets to invalidate cache, but that means we can't provide users with a stable URL to `curl`.

Since this part of the doc is considered deprecated, and will be removed in the future, I just reverted them to nomadproject.io.

File: website/content/docs/integrations/consul/index.mdx [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+  the Consul default.
+
+* job run command: <EnterpriseAlert inline/> If the
+  [`-consul-namespace`][consul_run_namespace] command line argument is specified
+  on job submission, this namespace will take precedence over the namespace set
+  in Nomad agent configuration.
+
+* group and task configuration: <EnterpriseAlert inline/> If the group level or
+  task level Consul [namespace field][consul_jobspec_namespace] is configured,
+  this namespace will take precedence over all other options.
```
[the remark this answers is not in the document]
Comment (lgfa29) [reply website/content/docs/integrations/consul/index.mdx:107]: Good point. I actually didn't even read this part, just copy-and-pasted from the original source 😅 

Re-order and re-worded so it sounds less if-else conditional. Also added the `CONSUL_NAMESPACE` env var since it's not mentioned in the `-consul-namespace` flag docs.

image

File: website/content/docs/integrations/vault/acl.mdx [hunk 15]
```diff
@@ -1,137 +1,257 @@
...
-other secrets. To enable secure, auditable and easy access to your secrets,
-Nomad integrates with HashiCorp's [Vault][]. Nomad servers and clients
-coordinate with Vault to derive a Vault token that has access to only the Vault
-policies the tasks needs. Nomad clients make the token available to the task and
-handle the tokens renewal. Further, Nomad's [`template` block][template] can
-retrieve secrets from Vault making it easier than ever to secure your
-infrastructure.
+The Vault ACL system protects the cluster from unauthorized access. When
+enabled, both Vault and Nomad must be properly configured in order for their
+integrations to work.
```
[the remark this answers is not in the document]
Comment (lgfa29) [reply website/content/docs/integrations/vault/acl.mdx:11]: Ah nice catch! Yeah, it's copy-pasta everywhere. I may have it like this in the tutorial as well 😅 

Updated to:

> The Vault ACL system protects the cluster from unauthorized access. It must be
properly configured in order for the Vault and Nomad integrations to work.

File: website/content/docs/integrations/vault/acl.mdx [hunk 16]
```diff
@@ -1,137 +1,257 @@
...
 
-## Agent Configuration
+Starting in Nomad 1.7, Nomad clients can use a task's [Workload Identity][] to
+authenticate to Vault and obtain a Vault ACL token specific to the task. When
+using Nomad workload identities, you no longer need to pass in a Vault ACL
+token to submit a job.
 
-To enable Vault integration, please see the [Nomad agent Vault
-integration][config] configuration.
+By default, Nomad only generates a workload identity for tasks that is used to
```
[the remark this answers is not in the document]
Comment (lgfa29) [reply website/content/docs/integrations/vault/acl.mdx:20]: Updated in Consul as well.

File: website/content/docs/integrations/consul/acl.mdx [hunk 17]
```diff
@@ -0,0 +1,446 @@
...
+must be configured to accept workload identities from tasks and services
+running in Nomad.
+
+## Nomad Agents
+
+Nomad agents need access to Consul in order to register themselves in the
+service catalog and discover other Nomad agents via service discovery for
+automatic clustering. Nomad servers also create [configuration
+entries][consul_config_entry] for Consul Service Mesh, so the specific
+permissions vary slightly between Nomad servers and clients.
```
[the remark this answers is not in the document]
Comment (lgfa29) [reply website/content/docs/integrations/consul/acl.mdx:23]: Update to this. I wanted to avoid giving too much instructions here and present this more as a documentation of what Nomad needs to work. But happy to reword it the suggested way as well.

image

[the remark this answers is not in the document]
Comment (tgross) [reply website/content/docs/integrations/consul/acl.mdx:23]: Perfect!

File: website/content/docs/integrations/consul/index.mdx (hunk 14 above)
[the remark this answers is not in the document]
Comment (tgross) [reply website/content/docs/integrations/consul/index.mdx:107]: That works!
